File: StringCrypter.swift
// StringCrypter.swift
import Foundation

let STATIC_KEY_SWIFT = "test_key"
let SALT_LENGTH_BYTES_SWIFT = 16 // 16 bytes = 32 hex characters

// Helper to convert Data to Hex String
extension Data {
    func hexEncodedString() -> String {
        return map { String(format: "%02hhx", $0) }.joined()
    }
}

// Helper to convert Hex String to Data
extension String {
    func hexDecodedData() -> Data? {
        var data = Data(capacity: self.count / 2)
        var index = self.startIndex
        while index < self.endIndex {
            let nextIndex = self.index(index, offsetBy: 2)
            guard nextIndex <= self.endIndex,
                  let b = UInt8(self[index..<nextIndex], radix: 16) else {
                // print("Error: Non-hex character found or odd-length string")
                return nil // Or handle error appropriately
            }
            data.append(b)
            index = nextIndex
        }
        return data
    }
}


func generateSaltSwift() -> String {
    var saltBytes = Data(count: SALT_LENGTH_BYTES_SWIFT)
    _ = saltBytes.withUnsafeMutableBytes { mutableBytes in
        SecRandomCopyBytes(kSecRandomDefault, SALT_LENGTH_BYTES_SWIFT, mutableBytes.baseAddress!)
    }
    return saltBytes.hexEncodedString()
}

func xorDataSwift(input: Data, key: Data) -> Data {
    var output = Data(count: input.count)
    guard !key.isEmpty else { return input }
    for i in 0..<input.count {
        output[i] = input[i] ^ key[i % key.count]
    }
    return output
}

func encryptSwift(plaintext: String, keyString: String) -> String? {
    let saltHex = generateSaltSwift()
    guard let plaintextData = (saltHex + plaintext).data(using: .utf8),
          let keyData = keyString.data(using: .utf8) else {
        return nil
    }

    let xoredData = xorDataSwift(input: plaintextData, key: keyData)
    return xoredData.base64EncodedString()
}

func decryptSwift(encryptedData: String, keyString: String) -> String? {
    guard let decodedData = Data(base64Encoded: encryptedData),
          let keyData = keyString.data(using: .utf8) else {
        // print("Error: Invalid Base64 string or key string.")
        return nil
    }

    let decryptedWithSaltData = xorDataSwift(input: decodedData, key: keyData)
    guard let decryptedWithSaltString = String(data: decryptedWithSaltData, encoding: .utf8) else {
        // print("Error: Could not convert decrypted data to UTF-8 string.")
        return nil
    }

    let saltHexLen = SALT_LENGTH_BYTES_SWIFT * 2
    guard decryptedWithSaltString.count >= saltHexLen else {
        // print("Error: Decrypted data too short to contain salt.")
        return nil
    }
    let saltEndIndex = decryptedWithSaltString.index(decryptedWithSaltString.startIndex, offsetBy: saltHexLen)
    return String(decryptedWithSaltString[saltEndIndex...])
}

func processStringSwift(type: Character, data: String) -> String? {
    let key = STATIC_KEY_SWIFT
    switch type {
    case "e":
        return encryptSwift(plaintext: data, keyString: key)
    case "d":
        return decryptSwift(encryptedData: data, keyString: key)
    default:
        print("Error: Invalid type specified. Use 'e' for encrypt or 'd' for decrypt.")
        return nil
    }
}

// Main test area
func runSwiftTests() {
    print("--- Swift String Encryption/Decryption Tests ---")

    let originalText = "Hello from Swift!"
    print("Original Text: \(originalText)")

    if let encryptedText = processStringSwift(type: "e", data: originalText) {
        print("Encrypted (Swift): \(encryptedText)")

        if let decryptedText = processStringSwift(type: "d", data: encryptedText) {
            print("Decrypted (Swift): \(decryptedText)")
            if decryptedText == originalText {
                print("Swift Encryption/Decryption Test: SUCCESSFUL")
            } else {
                print("Swift Encryption/Decryption Test: FAILED")
                print("Expected: \(originalText), Got: \(decryptedText)")
            }
        } else {
            print("Swift Decryption FAILED (result was nil).")
        }
    } else {
        print("Swift Encryption FAILED (result was nil).")
    }

    print("\n--- Interoperability Test (Swift decrypts Python) ---")
    // Example: String "Hello from Python for Swift!" encrypted by Python
    let pythonEncrypted = "NzM2MzMxMzEzNjY0NjEzMDYxMzQzNjY0NjIzNjYyMzQzNFN1Z2RjY2RjZWNkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2Q=" // Placeholder
    let expectedPythonDecryption = "Hello from Python for Swift!" // Placeholder

    print("Python Encrypted: \(pythonEncrypted)")
    if let decryptedFromPython = processStringSwift(type: "d", data: pythonEncrypted) {
        print("Decrypted by Swift: \(decryptedFromPython)")
        if decryptedFromPython == expectedPythonDecryption {
            print("Swift decryption of Python string: SUCCESSFUL (with correct Python output)")
        } else {
            print("Swift decryption of Python string: FAILED or placeholder data used.")
            print("Expected: \(expectedPythonDecryption), Got: \(decryptedFromPython)")
        }
    } else {
        print("Swift decryption of Python string FAILED (result was nil).")
    }
     print("Note: For the Python interop test to be meaningful, replace 'pythonEncrypted' with actual output from your Python script for the string '\(expectedPythonDecryption)'.")


    print("\n--- Interoperability Test (Swift encrypts for others) ---")
    let swiftMsgForOthers = "Hello from Swift for other languages!"
    if let swiftEncryptedForOthers = processStringSwift(type: "e", data: swiftMsgForOthers) {
        print("Swift Encrypted for others (Original: '\(swiftMsgForOthers)'): \(swiftEncryptedForOthers)")
        print("Take this string and try to decrypt it using process_string('d', ...) in other languages.")
    } else {
        print("Swift encryption for others FAILED (result was nil).")
    }
}

// To run these tests, you would typically call runSwiftTests()
// In a Swift playground, the output will appear automatically.
// For a command-line tool, you'd have a main.swift that calls this.
 runSwiftTests()

File: string_crypter.pl
# string_crypter.pl
use strict;
use warnings;
use MIME::Base64 qw(encode_base64 decode_base64);
use Encode qw(encode_utf8 decode_utf8); # For explicit UTF-8 handling

# For more cryptographically secure random bytes, one might use:
# use Bytes::Random::Secure qw(random_bytes); # If available
# For this example, we'll use a simpler approach for salt.

our $STATIC_KEY_PL = "test_key";
our $SALT_LENGTH_BYTES_PL = 16; # 16 bytes = 32 hex characters

sub generate_salt_pl {
    my $salt_bytes = "";
    for (1..$SALT_LENGTH_BYTES_PL) {
        $salt_bytes .= chr(int(rand(256)));
    }
    # Convert bytes to hex string
    return unpack('H*', $salt_bytes);
}

sub xor_strings_pl {
    my ($s1, $s2_key) = @_;
    return $s1 if length($s2_key) == 0;

    my $s1_len = length($s1);
    my $s2_len = length($s2_key);
    my $result = "";

    for (my $i = 0; $i < $s1_len; $i++) {
        $result .= chr(ord(substr($s1, $i, 1)) ^ ord(substr($s2_key, $i % $s2_len, 1)));
    }
    return $result;
}

sub encrypt_pl {
    my ($plaintext, $key_string) = @_;
    my $salt_hex = generate_salt_pl();
    
    # Prepend raw salt bytes (converted from hex) to UTF-8 encoded plaintext
    my $salt_raw = pack('H*', $salt_hex);
    my $plaintext_utf8 = encode_utf8($plaintext); # Ensure plaintext is UTF-8
    my $data_with_salt = $salt_raw . $plaintext_utf8;

    my $key_bytes = encode_utf8($key_string); # Ensure key is UTF-8

    my $xored_data = xor_strings_pl($data_with_salt, $key_bytes);
    my $encoded_base64 = encode_base64($xored_data, ""); # "" removes newlines
    return $encoded_base64;
}

sub decrypt_pl {
    my ($encrypted_data, $key_string) = @_;
    my $decoded_base64;
    eval {
        $decoded_base64 = decode_base64($encrypted_data);
    };
    if ($@) {
        print "Decryption error: Invalid Base64 string. $@\n";
        return undef;
    }

    my $key_bytes = encode_utf8($key_string); # Ensure key is UTF-8

    my $decrypted_with_salt = xor_strings_pl($decoded_base64, $key_bytes);

    my $salt_raw_len = $SALT_LENGTH_BYTES_PL; # Raw salt length
    if (length($decrypted_with_salt) < $salt_raw_len) {
        print "Decryption error: Decrypted data is too short to contain salt.\n";
        return undef;
    }

    # Salt is raw bytes, plaintext part needs UTF-8 decoding
    # my $salt_part = substr($decrypted_with_salt, 0, $salt_raw_len); # Not needed for output
    my $plaintext_part_utf8 = substr($decrypted_with_salt, $salt_raw_len);
    
    my $original_plaintext;
    eval {
        $original_plaintext = decode_utf8($plaintext_part_utf8, Encode::FB_CROAK);
    };
    if ($@) {
        print "Decryption error: Could not decode UTF-8 from decrypted data. $@\n";
        return undef; # Or handle error, e.g. return raw bytes
    }
    
    return $original_plaintext;
}

sub process_string_pl {
    my ($type, $data) = @_;
    my $key = $STATIC_KEY_PL;
    if ($type eq 'e') {
        return encrypt_pl($data, $key);
    } elsif ($type eq 'd') {
        return decrypt_pl($data, $key);
    } else {
        print "Error: Invalid type specified. Use 'e' for encrypt or 'd' for decrypt.\n";
        return undef;
    }
}

# Main test area
if ($0 eq __FILE__) {
    print "--- Perl String Encryption/Decryption Tests ---\n";

    my $original_text = "Hello from Perl!";
    print "Original Text: $original_text\n";

    my $encrypted_text = process_string_pl('e', $original_text);
    if (defined $encrypted_text) {
        print "Encrypted (Perl): $encrypted_text\n";

        my $decrypted_text = process_string_pl('d', $encrypted_text);
        if (defined $decrypted_text) {
            print "Decrypted (Perl): $decrypted_text\n";
            if ($decrypted_text eq $original_text) {
                print "Perl Encryption/Decryption Test: SUCCESSFUL\n";
            } else {
                print "Perl Encryption/Decryption Test: FAILED\n";
                print "Expected: $original_text, Got: $decrypted_text\n";
            }
        } else {
            print "Perl Decryption FAILED (result was undef).\n";
        }
    } else {
        print "Perl Encryption FAILED (result was undef).\n";
    }

    print "\n--- Interoperability Test (Perl decrypts Python) ---\
";
    # Example: String "Hello from Python for Perl!" encrypted by Python
    my $python_encrypted = "NzM2MzMxMzEzNjY0NjEzMDYxMzQzNjY0NjIzNjYyMzQzNFN1Z2RjY2RjZWNkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2RjZWRlY2VjZGNlZGVjZWNkY2VkZWNlY2Q="; # Placeholder
    my $expected_python_decryption = "Hello from Python for Perl!"; # Placeholder

    print "Python Encrypted: $python_encrypted\n";
    my $decrypted_from_python = process_string_pl('d', $python_encrypted);
    if (defined $decrypted_from_python) {
        print "Decrypted by Perl: $decrypted_from_python\n";
        if ($decrypted_from_python eq $expected_python_decryption) {
            print "Perl decryption of Python string: SUCCESSFUL (with correct Python output)\n";
        } else {
            print "Perl decryption of Python string: FAILED or placeholder data used.\n";
            print "Expected: $expected_python_decryption, Got: $decrypted_from_python\n";
        }
    } else {
        print "Perl decryption of Python string FAILED (result was undef).\n";
    }
    print "Note: For the Python interop test to be meaningful, replace 'python_encrypted' with actual output from your Python script for the string '$expected_python_decryption'.\n";


    print "\n--- Interoperability Test (Perl encrypts for others) ---\n";
    my $perl_msg_for_others = "Hello from Perl for other languages!";
    my $perl_encrypted_for_others = process_string_pl('e', $perl_msg_for_others);
    if (defined $perl_encrypted_for_others) {
        print "Perl Encrypted for others (Original: '$perl_msg_for_others'): $perl_encrypted_for_others\n";
        print "Take this string and try to decrypt it using process_string('d', ...) in other languages.\n";
    } else {
        print "Perl encryption for others FAILED (result was undef).\n";
    }
}

1; # Required for 'use' and 'require'
